### PR TITLE
Testing full dichotomy solver in probabilistic frontend and marg query tests

### PR DESCRIPTION
**Pull request by choucavalier**

### CODE REVIEW

File: neurolang/probabilistic/dichotomy_theorem_based_solver.py [hunk 1]
```diff
@@ -235,53 +236,4 @@ def _maybe_reintroduce_head_variables(ra_query, flat_query, unified_query):
...
-    cpl_program : CPLogicProgram
-        CP-Logic program on which the query should be solved.
-
-    Returns
-    -------
-    ProvenanceAlgebraSet
-        Provenance set labelled with probabilities for each tuple in the result
-        set.
-
-    """
```
Comment (demianw) [inline neurolang/probabilistic/dichotomy_theorem_based_solver.py:255]: The documentation shouldn't have been eliminated

File: neurolang/probabilistic/dalvi_suciu_lift.py [hunk 2]
```diff
@@ -163,7 +170,11 @@ def dalvi_suciu_lift(rule, symbol_table):
             for atom in extract_logic_atoms(rule)
         )
     ):
-        return TranslateToNamedRA().walk(rule)
+        free_vars = extract_logic_free_variables(rule)
+        rule = MakeExistentialsImplicit().walk(rule)
+        result = TranslateToNamedRA().walk(rule)
+        proj_cols = tuple(Constant(ColumnStr(v.name)) for v in free_vars)
+        return Projection(result, proj_cols)
```
Comment (demianw) [inline neurolang/probabilistic/dalvi_suciu_lift.py:191]: Isn't this covered by line 164? `convert_rule_to_ucq`

File: neurolang/probabilistic/tests/test_marg_query_resolution.py [hunk 3]
```diff
@@ -4,7 +4,11 @@
 from ...logic import Conjunction, Implication
 from ...relational_algebra import ColumnStr, NamedRelationalAlgebraFrozenSet
 from ...relational_algebra_provenance import ProvenanceAlgebraSet
-from .. import dichotomy_theorem_based_solver, weighted_model_counting
+from .. import (
+    dalvi_suciu_lift,
+    dichotomy_theorem_based_solver,
```
Comment (demianw) [inline neurolang/probabilistic/tests/test_marg_query_resolution.py]: maybe this should be renamed to `small_dichotomy_theorem_based_solver`

File: neurolang/probabilistic/tests/test_marg_query_resolution.py [hunk 4]
```diff
@@ -28,8 +32,14 @@
...
-    params=((weighted_model_counting, dichotomy_theorem_based_solver)),
-    ids=["SDD-WMC", "dichotomy-Safe query"],
+    params=(
+        (
+            weighted_model_counting,
+            dichotomy_theorem_based_solver,
+            dalvi_suciu_lift,
+        )
+    ),
+    ids=["SDD-WMC", "dichotomy-Safe query", "dalvi_suciu_lift"],
```
Comment (demianw) [inline neurolang/probabilistic/tests/test_marg_query_resolution.py:42]: what is "dichotomy-safe"?

**Review by demianw — CHANGES_REQUESTED** [review]

File: neurolang/probabilistic/tests/test_marg_query_resolution.py (hunk 4 above)
Comment (choucavalier) [reply neurolang/probabilistic/tests/test_marg_query_resolution.py:42]: I don't know, I think you initially wrote that, no? I'm renaming this to `small-dichotomy`

File: neurolang/probabilistic/dalvi_suciu_lift.py (hunk 2 above)
Comment (choucavalier) [reply neurolang/probabilistic/dalvi_suciu_lift.py:191]: `convert_rule_to_ucq` keeps existential quantifiers, resulting in the variable `rule` being passed to `TranslateToNamedRA` with existential quantifiers but `TranslateToNamedRA` does not translate existential predicates. I've added a test that fails here: https://github.com/tgy/NeuroLang/blob/lqp-testing/neurolang/probabilistic/tests/test_dalvi_suciu_lift.py#L489 

What do you think is the right solution?

File: neurolang/probabilistic/dichotomy_theorem_based_solver.py (hunk 1 above)
Comment (choucavalier) [reply neurolang/probabilistic/dichotomy_theorem_based_solver.py:255]: done

File: neurolang/probabilistic/tests/test_marg_query_resolution.py (hunk 3 above)
Comment (choucavalier) [reply neurolang/probabilistic/tests/test_marg_query_resolution.py]: done

File: neurolang/probabilistic/tests/test_marg_query_resolution.py (hunk 4 above)
Comment (choucavalier) [reply neurolang/probabilistic/tests/test_marg_query_resolution.py:42]: done

File: neurolang/probabilistic/tests/test_marg_query_resolution.py [hunk 5]
```diff
@@ -125,7 +125,7 @@ def test_marg_query_two_vars_conditioning(solver):
 
 
 def test_marg_query_conjunctive_conditioned_and_conditioning(solver):
-    if solver != dichotomy_theorem_based_solver:
+    if solver != small_dichotomy_theorem_based_solver:
```
Comment (demianw) [inline neurolang/probabilistic/tests/test_marg_query_resolution.py]: This shouldn't be skipped an exception should be returned and verified. There's examples of this in other cases

File: neurolang/probabilistic/tests/test_probabilistic_solvers.py [hunk 6]
```diff
@@ -693,7 +693,7 @@ def test_program_with_variable_equality(solver):
 
 
 def test_repeated_variable_probabilistic_rule(solver):
-    if solver is not dichotomy_theorem_based_solver:
+    if solver is not small_dichotomy_theorem_based_solver:
         pytest.skip()
```
Comment (demianw) [inline neurolang/probabilistic/tests/test_probabilistic_solvers.py]: same here this shouldn't be a skip but a check for the proper exception

File: neurolang/probabilistic/tests/test_probabilistic_solvers.py [hunk 7]
```diff
@@ -707,7 +707,7 @@ def test_repeated_variable_probabilistic_rule(solver):
 
 
 def test_repeated_variable_with_constant_in_head(solver):
-    if solver is not dichotomy_theorem_based_solver:
+    if solver is not small_dichotomy_theorem_based_solver:
         pytest.skip()
```
Comment (demianw) [inline neurolang/probabilistic/tests/test_probabilistic_solvers.py]: and here and every other one that is pertinent. Skipping the test really breeds for bugs

**Review by demianw — CHANGES_REQUESTED** [review]
Last comments

File: neurolang/probabilistic/containment.py [hunk 8]
```diff
@@ -40,8 +40,8 @@ def freeze_atom(atom):
     the symbol.
     '''
     args = (
-        Constant(s.name)
-        for s in atom.args
+        Constant(arg.name) if isinstance(arg, Symbol) else arg
+        for arg in atom.args
```
Comment (choucavalier) [inline neurolang/probabilistic/containment.py:44]: Really not sure about this. This code was failing when the atom contained `Constant` terms, as the function assumed all terms to be `Symbol`s

File: neurolang/probabilistic/dalvi_suciu_lift.py [hunk 9]
```diff
@@ -135,6 +142,12 @@ def solve_succ_query(query, cpl_program):
                 "Query %s not liftable, algorithm can't be applied",
                 query
             )
+        # project on query's head variables
+        ra_query = _project_on_query_head(ra_query, shattered_query)
+        # re-introduce head variables potentially removed by unification
+        ra_query = _maybe_reintroduce_head_variables(
+            ra_query, flat_query, unified_query
+        )
```
Comment (choucavalier) [inline neurolang/probabilistic/dalvi_suciu_lift.py:154]: This is done to make sure the answered tuples match the head of the rule (similarly to what we did in the small dichotomy theorem). This fixes issues with head of queries containing constants for example

**Review by demianw — CHANGES_REQUESTED** [review]
@tgy looks great! Really good job. Just a little ask,
a coverage analysis shows (among others less important) that lines 316 and 327 in the following code (neurolang/probabilistic/dalvi_suciu_lift.py) are not covered. Could you please ask 2 small tests to cover them? Then we merge!
```python
307 def extract_probabilistic_root_variables(formulas, symbol_table): 
308    candidates = None 
309    for formula in formulas: 
310        probabilistic_atoms = OrderedSet( 
311            atom 
312            for atom in extract_logic_atoms(formula) 
313            if not is_atom_a_deterministic_relation(atom, symbol_table) 
314        ) 
315        if len(probabilistic_atoms) == 0: 315 ↛ 316
316            continue 
317        root_variables = reduce( 
318            lambda y, x: set(x.args) & y, 
319            probabilistic_atoms[1:], 
320            set(probabilistic_atoms[0].args) 
321        ) 
322        if candidates is None: 
323            candidates = root_variables 
324        else: 
325            candidates &= root_variables 
326    if candidates is None: 326 ↛ 327
327        candidates = set() 
328    return candidates 
```

**Comment by demianw** [comment]
Great job @tgy

**Comment by choucavalier** [comment]
Thanks for the great reviews, as always :-)

**Pull request merged**